### ISSUE TITLE
Hardware acceleration on Apple Silicon with Metal plugin

### ISSUE THREAD

**Issue by dannote** [google/jax#8074]
Hello!

I'm looking for a way to accelerate the XLA compiler for Apple's M1 (elixir-nx/nx#490). Apple provides a PluggableDevice plugin with the `METAL` platform, but it doesn't include an XLA backend for it yet.

Do you have any plans to target M1's GPU?

**Comment by nicholasjng** [comment]
Just out of curiosity, how would one go about implementing a GPU backend for Metal in XLA? I checked out the XLA source a little bit, and while it looks fairly complicated (though I also have not worked with C++ in a while), do you think it could be a feasible project for someone to implement some GPU-accelerated XLA operations for the M1 platform? (I've been looking to get into GPU programming for a while now, which is the reason for my interest).

If the hypothetical path from IREE to Metal you mentioned is much more realistic, though, then it might not make sense to go ahead on this. I would be happy about a quick expert opinion :)

**Comment by hawkinsp** [comment]
Well, the path to get something working is perhaps shorter than you might think:

* JAX now has an experimental IREE backend, which is quite incomplete, but might be an excellent starting point for someone who wanted to look at Metal support via IREE. To use it, you simply need to do this:

```
pip install --upgrade jaxlib
pip install iree-compiler-snapshot iree-runtime-snapshot -f https://github.com/google/iree/releases
JAX_ENABLE_MLIR=1 JAX_PLATFORMS=iree python your_jax_script.py
```
So the JAX->IREE path exists, at least in an early form.

* IREE has had a Metal backend in the past, but apparently it needs to be revived: https://github.com/google/iree/issues/4370

So if I wanted to try this, I might try looking at that IREE issue.

**Comment by nicholasjng** [comment]
Thanks for the hint! I need to take a closer look at IREE, then. Seems they want it rewritten in Objective-C for better performance and smaller size, so I will check up on how they implement other C backends first.

**Comment by benjaminpope** [comment]
This would be awesome!

**Comment by nicholasjng** [comment]
I looked at IREE a little bit, and it seems out of my league (at least for the moment and given my free time budget), so I wanted to ask some follow-up questions about developing a new XLA backend in the case of Metal (if that's ok):

* I looked at the docs on the Tensorflow XLA page, and understood that you need to do less/more work depending on whether an LLVM backend exists for your target platform. As far as I could tell, there actually exists an LLVM-based Metal compiler in the XCode toolchain (`xcrun metal`), but Apple has not open-sourced it. Could you maybe use that to programmatically feed it JAX HLO and get Metal shaders out?
* Apple open-sourced a [C++ wrapper](https://developer.apple.com/metal/cpp/) for Metal a few months ago, can that be used to build an XLA extension in C++ for dispatching Metal shaders?

Apologies if this is obvious, I'm not really experienced with compilers (to my great regret, it's a very interesting subject to me). I want to progress on this, though, but right now the barrier is a little too steep. Thank you for your consideration!

**Comment by stellaraccident** [comment]
FYI - news from one of IREE's users/contributors relevant to this: https://nod.ai/pytorch-m1-max-gpu/

They were successful at adapting and tuning IREE to the case. This is still quite early work but promising.

**Comment by powderluv** [comment]
yes JAX+IREE (and nod.ai tuned SHARK) on Apple Silicon is a high priority for us at Nod.ai. We hope to have the JAX pipeline flushed out once a few more upstream pieces land. But happy to help with anything if anyone else is trying it.

**Comment by powderluv** [comment]
@hawkinsp the Build from source step shouldn't be required any longer. The nightly builds with VULKAN turned on.  Please let me know if that is not the case - we have already upstreamed all the apple M1 pieces. 

You will need to pass "extra_args" to IREE to set the right  Apple M1 target triple: 
https://github.com/nod-ai/SHARK/blob/1186d7c58e6046aea6a6115c608dbd77728e7aca/shark/iree_utils.py#L93-L96

If you have a nice JAX model you would like optimized for the M1 Ultra please send me link and we can tweak the backend codegen for that.

**Comment by hawkinsp** [comment]
Thanks! You're right, everything needed is in an `iree` nightly build. And passing the right target triple meant I got meaningfully better performance for the `f32[1000, 1000]` x `f32[1000, 1000]` matmul via the Vulkan path, so that's pretty promising.

If you're interesting in performance tuning, I'm sure we can cook up a transformer model or something that might be fun to optimize!

**Comment by Birch-san** [comment]
thanks for everyone's effort so far in getting jax to work with M1 GPU.

I tried this out today on the [dalle-playground](https://github.com/saharmor/dalle-playground) jax model, but I got an error:  
`failed to legalize operation 'mhlo.scatter' that was explicitly marked illegal` _([full details here](https://gist.github.com/Birch-san/fb3d77f102bb177e4095df5c9be5702f))_.

is this currently unimplemented? is there an option I can use to lower the operations differently?  
_I didn't see any other options in [`Flow/Transforms/Passes.cpp`](https://github.com/google/iree/blob/main/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp) or [`MHLO/Passes.cpp`](https://github.com/google/iree/blob/main/compiler/src/iree/compiler/InputConversion/MHLO/Passes.cpp) that looked like capable of fixing this._

I _assume_ I installed everything correctly.

```bash
# on commit 345cc19949273cc414d94e6f13d0620b780af465
git clone https://github.com/google/jax.git
cd jax
# build jaxlib wheel
/Users/birch/anaconda3/envs/torch-nightly/bin/python ./build/build.py \                                               
--noenable_mkl_dnn \
--noenable_cuda \
--noenable_tpu \
--noenable_nccl
# install jaxlib wheel
pip install /Users/birch/git/jax/dist/jaxlib-0.3.11-cp39-none-macosx_11_0_arm64.whl
# install jax
pip install -e .
# install today's iree release candidate
pip install iree_compiler iree_runtime -f https://github.com/google/iree/releases/tag/candidate-20220606.161
```

@powderluv if you're looking for a model to optimize the backend against, this [`dalle-playground`](https://github.com/saharmor/dalle-playground) is pretty popular right now (powers the [`dalle-mini`](https://huggingface.co/spaces/dalle-mini/dalle-mini) demo, but also has capability to [run `dalle-mega`](https://github.com/saharmor/dalle-playground/blob/5a7be250e5c377b18a3f9f07631d177bcf91c934/backend/app.py#L37), which is not yet featured in the demo).

**Comment by powderluv** [comment]
oh this looks exciting. Can you please share your entire command history after installing the iree_compiler/runtime ?  Can I run it from the command line instead of VScode ?

**Comment by Birch-san** [comment]
yeah, definitely possible to run without VSCode.

my command history for setting up [`dalle-playground`](https://github.com/saharmor/dalle-playground) was pretty messy. broadly speaking it can be as simple as this:

## Setup Anaconda

Install Anaconda via [command-line arm64 installer](https://www.anaconda.com/products/distribution), then:

### (optional) install PyTorch nightly (to get M1 GPU support)

```bash
conda update -n base -c defaults conda
conda create -n torch-nightly python=3.9
pip install --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/cpu
conda config --set auto_activate_base false
echo 'conda activate torch-nightly' >> ~/.zshrc
```

## Setup [`dalle-playground`](https://github.com/saharmor/dalle-playground)

### Backend

#### Install

```bash
git clone https://github.com/saharmor/dalle-playground.git
cd dalle-playground
# ensure you have on your PATH: cargo + a beta release of rustc
pip install -r requirements.txt
```

Beware: the `pip install -r requirements.txt` was a bit challenging; one of its dependencies tried to compile-from-source.  
It wanted `rustc` and `cargo` to be on the `PATH`.  
Stable `rustc` wasn't new enough; I had to install a beta. These worked:

```
rustc 1.62.0-beta.3 (a5cf77ca6 2022-06-02)
cargo 1.62.0-beta.3 (4751950cc 2022-05-27)
```

If you're using macOS with nix as your package manager, do tell and I can share the config I used to install Rust. It was horrible.

#### Run

```bash
python app.py 8080
```

### Frontend (easy-peasy, NodeJS stuff)

```bash
cd frontend
npm i
npm start
```

Ensure your backend is ready (i.e. wait for it to have printed "--> DALL-E Server is up and running!"), then navigate to:  
http://localhost:3000/dalle-playground?backendUrl=http://127.0.0.1:8080

Type a query (this one succeeded because I ran without GPU):

<img width="811" alt="image" src="https://user-images.githubusercontent.com/6141784/172262546-d4e256d4-30dc-4776-a873-053283645a06.png">

**Comment by Birch-san** [comment]
I've filed an [issue with iree](https://github.com/google/iree/issues/9361) to improve support for the `mhlo.scatter` operation used in this jax model.

in the meantime: is there a way to utilise more of the _CPU_? I think I'm noticing the same thing as https://github.com/google/jax/issues/5022 — when I ask the jax model to generate 1 image on-CPU, it takes a few minutes (2mins with dalle-mini, 27mins with dalle-mega) but my CPU usage is only 117%. moreover it seems to be utilising efficiency cores without maxing out performance cores.  
is there a way to set thread affinity so that it prefers to run on performance cores? and a way to tell it to use _all_ of them?

**Comment by hawkinsp** [comment]
@Birch-san Yes, the MHLO op support in IREE isn't complete and has some known deficiencies. see for example the hotlist of issues: https://github.com/google/iree/issues?q=is%3Aissue+is%3Aopen+label%3Aintegrations%2Fmhlo many of which come from running the JAX tests. But it's early days yet! I'm sure the IREE folks would appreciate you filing issues as you discover them, assuming they aren't already known.

For using more CPU cores: XLA doesn't do a whole lot of inter-operation parallelism on CPU at the moment. You can add explicit parallelism at the JAX level (e.g., using `pmap` or `pjit`). I'm not sure what IREE-on-CPU's parallelism status is. CPU has been a bit neglected historically in XLA but there is a bunch of work in progress from a number of compiler folks that gives me hope that it will start to make more rapid progress soon.

**Comment by Birch-san** [comment]
yeah, happy to file any issues I find 🙂 but I can't get past that first one, so I think that's all I can report for now on the IREE side.

okay, I suppose pursuing parallelism on the JAX level would require me to get deep into rewriting the model (not my specialty).  
I have spare CPU cores, so I can run multiple tasks concurrently. with dalle-mega, I can generate 1 image on-CPU in 27 mins (117% CPU usage), or submit 4 concurrent 1-image tasks (450% CPU) to get 4 images in 43 mins (11 mins per pic).  
a good improvement, but would be so nice to get the GPU involved too!

**Comment by Birch-san** [comment]
sure, I'll file an issue about CPU parallelism too. thanks for your theories. 🙂

**Comment by Birch-san** [comment]
@stellaraccident I've raised https://github.com/google/iree/issues/9368 to report the "1 CPU core" problem.

thanks for explaining how to pass flags.

I tried placing this at the start of `app.py`:

```python
from iree import runtime as rt
rt.flags.parse_flags("--task_topology_group_count=8")
```

but it didn't make any difference; CPU usage for generating 1 image in dalle-mini was 125% either way, and took 139secs either way.

is there a breakpoint I can place anywhere in the iree python to determine what group count it ends up using? or a Python expression I can add?

**Comment by hawkinsp** [comment]
@Birch-san Can you confirm that you're actually running the IREE backend (`JAX_PLATFORMS=iree` or similar) when measuring performance? Note you will probably hit the `mhlo.scatter` problem you reported above.

**Comment by Birch-san** [comment]
@hawkinsp oh, that's embarrassing.  
you're right -- I was **not** using `JAX_PLATFORMS=iree` for my on-CPU test. and as you say: if I _do_ turn it on, it explodes with the `mhlo.scatter` complaint.
  
presumably that explains why `--task_topology_group_count=8` had no effect.

@stellaraccident I guess I'll need to update the iree issue. it sounds like it's still an issue that ought to exist, but I think my repro via dalle-playground is not valid, and the detail about `task_topology_group_count` not working is also invalid.

**Comment by stellaraccident** [comment]
We're happy to take the issue: this should be more automatic.

But this is what we need to focus on fixing to unblock: https://github.com/google/iree/issues/9361

**Comment by Birch-san** [comment]
I tried the proposed patch for https://github.com/google/iree/issues/9361. that gets me onto the next error with using the IREE backend to run dalle-playground: https://github.com/google/jax/issues/11166 (`MLIR translation rule for primitive 'remat_call' not found for platform 'iree'`)

**Comment by philipturner** [comment]
I’m seeking to run AlphaFold, which relies on JAX, on the M1 Max GPU. People got it running on CPU outside of the Linux Docker container, but the only supported GPUs are NVIDIA. I have quite a bit of experience with Metal and recently researched transpiling LLVM IR into AIR, so I could contribute to a JAX Metal port.

A big concern is performance for matrix multiplication kernels. MPS used by PyTorch has has 3x faster performance than non-tuned matrix multiplication kernels, because it uses `simdgroup_matrix` hardware-accelerated instructions.

**Comment by Robokan** [comment]
Curious what the current status of this is. If I install JAX on a Mac M1 via:

pip install jax

Is it running on the GPU by default? Or do I need to do something to enable it?

**Comment by leonard-gleyzer** [comment]
Am also curious about this. JAX M1-series GPU support would be incredibly helpful.

**Comment by stellaraccident** [comment]
We're working on a better layered plugin architecture for Jax, and plugging iree in more directly is pushing that forward. This will make it easier for third parties to extend real support in this direction.

Despite early results, I do suspect that a real/solid/performant backend for the M1 GPU is quite a bit of work, and my goal is to enable the community to service this area of need based on some starting points that work. It would also be ideal to involve Apple themselves.

Stand by for a better plugin.

**Comment by nicholasjng** [comment]
I think the decision to pluck out XLA from the Tensorflow repo might spark some development in that regard. I have been trying to get into IREE and XLA after work for some time now, but the complexity is very high and thus there is a more or less substantial barrier of entry for a layperson like me.

There was some documentation in TF XLA on how to implement a new backend, but it was rather abstract. Maybe there could be some more in-depth documentation walkthrough on how to add new backends (or even frontends) in the future? I saw that MLIR has done some work in that regard with its toy dialect example. I think it would be a valuable thing to add - would love to get involved in it as well if that's feasible!

**Comment by stellaraccident** [comment]
Good feedback - thank you. We're rolling up the sidewalks here at the Google offices for the year, so to set expectations, this is likely a 2023 discussions. If there is demand, it would be good to get this on the OpenXLA calendar to talk through... That could be a good forum to connect more of the dots you are seeking and get more things flushed out publicly.

**Comment by stellaraccident** [comment]
@skye is there any chance of releasing more of the documentation/discussion about the plugin with you are doing? Maybe connect with Thea on the OpenXLA side to create a space for this kind of discussion?

**Comment by philipturner** [comment]
Just to note, the CPU on Apple silicon is very powerful. The AMX blocks provide up to [2 TFLOPS of single-precision matmul power](https://github.com/dougallj/applegpu/issues/21) on the M1 Max, and 1 TFLOPS on the base M1. Compare that to 8 TFLOPS for M1 Max GPU and 2 TFLOPS for base M1 GPU. If you don't fully optimize matrix multiplications for the GPU, it will be  equal slower than CPU. Hence, no point in GPU acceleration.

The Apple GPU architecture, by itself, it bad at matrix multiplications. A generic matmul library called DLPrimitives reached 25% ALU utilization, while pre-A14 MPSMatrixMultiplication also reached that. Compare that to A14/M1 and later, where MPSMatrixMultiplication jumped to 80% ALU utilization. Apple added a special intrinsic called `simdgroup_matrix` which allows fast matrix multiplications. We would need to expose this intrinsic to the SPIR-V backend compiler. Otherwise, 25% times 2.6 TFLOPS = 0.65 TFLOPS, which is less than 1.0 TFLOPS from M1 CPU.

>  It would also be ideal to involve Apple themselves.

From personal experience, I would not count on Apple helping us out. I also think this effort is mostly wasted anyway, because all the backend porting will become obsolete soon. It might be good to read over [Modular AI](https://www.modular.com/)'s website, then wait to integrate their platform into JAX in the future.

**Comment by stellaraccident** [comment]
This is what I was alluding to: without exposing the intrinsics and using them, there is really no way to great. I haven't checked recently but I generally assume that such things are inaccessible without the vendor having taken steps to make them so or worked jointly on the solution.

Well aware of modular. If they produce something better on the dimensions we care about, you certainly won't see me pushing back on leveraging it. I'm reserving my own judgment until there is more than marketing to evaluate.

**Comment by philipturner** [comment]
Your opinion about modular does make sense. Also, it will be difficult to, once you expose the intrinsics, ensure they're always used properly to reach >38% ALU utilization (1.0/2.6). You'd basically be rewriting `MPSGraph` from scratch, including Winograd convolutions and the likes. That's a massive time investment.

**Comment by stellaraccident** [comment]
It sometimes not that bad, but it is very hard to chart a path to great on that stuff without the hardware maker's involvement. These things are always hard to hold right -- and that is when you have a great datasheet, docs, etc.

**Comment by philipturner** [comment]
> on that stuff without the hardware maker's involvement.

And this is exactly why so many AI and HPC applications are vendor-locked to NVIDIA. Their employees actively assist with optimizing several frameworks for CUDA. In contrast, with AMD and Apple it's somewhat more "you're on your own" (HIPify and SYCLomatic seem designed to address this). The solution isn't to get 4x as many engineers (from NVIDIA + AMD + Intel + Apple) performing 4x the optimization.

**Comment by stellaraccident** [comment]
We've had substantially better experiences than "4x" with community support for AMD. There have been more developments than this which will hopefully be written up soon: https://nod.ai/shark-rdna3-sd/

It took about 6 weeks for AMD and a contractor to adapt code generation solutions to get to quite compelling performance (there's more actively discussed on discord that doesn't seem to have written official statements). The best experiences do come from vendor engagement, but ideally that involves last mile level porting vs redevelopment. As an example, AMD seems to have made extensions to their driver to enable cutting edge support to get there: https://www.amd.com/en/support/kb/release-notes/rn-rad-win-22-11-1-mlir-iree (ie. This seems to always be better as a partnership with the hardware vendor to ensure that needed facilities are exposed).

**Comment by philipturner** [comment]
AMD recently has been catching up to NVIDIA. Has the same effort also been puts forth with Intel’s Arc GPUs, and ARM’s Mali GPUs? As a more contrived example, there are AI processor startups with very small teams. They can only afford to invest time optimizing their hardware for (1) IREE, (2) PyTorch, (3) TensorFlow, (4) another less popular framework (e.g. [S4TF](https://github.com/s4tf)). If I were them, I would choose PyTorch and invest remaining person-hours into maintaining the PyTorch backend. That leaves other frameworks untouched.

> We've had substantially better experiences than "4x" with community support for AMD

I was not thinking 4x quality, rather 4x the amount of person-hours. You shouldn’t need to hire a contractor just to make your company’s hardware run optimally, on the newest popular frontend or IR.

**Comment by stellaraccident** [comment]
I think this all may be further along on these axes than you are thinking, but I can really only point to publicly released statements. These classes of parts are very much in the space of enhancements to one applying to all, especially those in the same broad API family -- which covers a lot at this point.

In my mind, this really is about API family + access to key intrinsics/extensions necessary for perf. Many of these are quite close on that category and get a lot for free. The further a part is on its own on that, the more expensive it will be and the higher the likelihood that only the vendor can do the work to get it there. Which brings us back to the original topic -- I suspect that Apple GPUs will be harder than expected to make great and probably require movement from Apple on programmability to get all the way. May be over estimating that but it's my default assumption for a closed platform like this.

**Comment by philipturner** [comment]
I will agree to disagree; your comment does have sound reasoning. I can cite evidence that Apple’s Metal team is extremely understaffed, and can barely maintain their own documentation or respond to developers. Choosing to work on IREE may not be in their interests.

Besides Apple, I may be one the few people with the capability to optimize this for the Apple GPU. Just check out the “Lumen & Nanite on MacOS” thread on Unreal Engine forums, or the failed [AlphaFold 2 port](https://github.com/philipturner/alphafold-metal), to see proof of this capability. However, I have other priorities so that leaves no one optimizing for Apple.

**Comment by powderluv** [comment]
This gives me a déjà vu of the discussion at

https://github.com/pytorch/pytorch/issues/47702#issuecomment-1018668822

But just to keep it to real code checked in (and not marketing hype or second guessing companies staffing etc) - we just contributed Winograd convs to IREE. It took us about 6 weeks. 

I hope we keep the discussion to technical implementations of the issue at hand.

**Comment by philipturner** [comment]
@powderluv I don't see how the comment you specifically linked is completely relevant to this thread. I do take it personally when someone exposes my own past mistakes, seemingly in order to discredit my argument. Would you mind editing the link to point to something more technical, such as [this comment](https://github.com/pytorch/pytorch/issues/47702#issuecomment-737682371), or remove it entirely? I can also refrain from expressing my viewpoint any further.

I don't assume you did this on purpose, but that PyTorch thread was extremely heated and a magnet for questionable behavior. There's a lot of hidden context that isn't particularly constructive to resurface. Best regards 😄

**Comment by philipturner** [comment]
> second guessing companies staffing etc)

Perhaps the proper context should be given: https://developer.apple.com/forums/thread/698037?answerId=731108022#731108022

> [@philipturner](https://developer.apple.com/forums/profile/philipturner) I think your frustration is justified. Apple's developer relations are unfortunately a mess. And it's not a criticism to the fine people who do all the hard work in the background and occasionally help us on these forums, but the obvious lack of structure and ownership in these matters. Lack of updates to the Metal Feature Set tables is just one symptom of a wide systemic problem. For example, the Metal Shading Language is very difficult to use as a reference tool due to subpar formatting and lack of hyperlinks. The API documentation is also lacklustre, incomplete and difficult to navigate. Forum communication is almost non-existent. It would be great if Apple considered creating a role dedicated to improving these aspects because it seems like this is something nobody really feels responsible for.
>
> Posted 2 months ago by  [jcookie](https://developer.apple.com/forums/profile/jcookie)

Another person noted some limitations faced in Apple-to-developer relations. I could have worded previous comments more politely, but I'm not trying to discredit Apple developers who work on Metal. My point was, these amazing people face real-world constraints. Enhancing IREE requires investing nonzero time and money, which could be spent elsewhere.

That's where my statement about Modular comes in. I am not an employee of them, and they do not endorse my viewpoint. They are in a very early stage, and they may not produce anything usable until several years. I suggested that Apple might view Modular, not IREE, as the solution to fragmentation in the AI industry. If they believed that, they might see contributing to IREE as effort spent in vain.

**Comment by powderluv** [comment]
Didn't mean to dig into anything from the past. Just want to say let's do what it takes to make JAX run well on Apple silicon. Thanks for being involved in the process.

**Comment by philipturner** [comment]
No worries. Thanks for being considerate.

**Comment by arpan-dhatt** [comment]
Instead of directly trying to produce Metal code for operations in XLA, what would be the viability of translating a jitted JAX call to an Metal Performance Shaders Graph invocation? MPS Graph has a number of features of XLA, although I'm not sure to what exact extent they overlap. This approach would have the benefit of using whichever intrinsics Apple's developers.

As I understand, there is currently no `PluggableDevice` support in JAX like TF, which would significantly increase the workload to see this task through.

Would this be a viable approach to get good GPU performance of JAX on Mac GPU's, ignoring how not using XLA might reduce the maximum possible performance? Are there any technical or design choices in JAX that would make such a task impossible (things like vmap, I'm not entirely sure)?

**Comment by mlaves** [comment]
FYI: There is a WWDC23 session namend "Optimize machine learning for Metal apps" that explicitly mentions "Discover the latest enhancements to accelerated ML training in Metal. Find out about updates to PyTorch and TensorFlow, and learn about __Metal acceleration for JAX__."

**Comment by PhilipVinc** [comment]
https://developer.apple.com/metal/jax/

**Comment by mlaves** [comment]
Was anyone able to get that running? I followed

> https://developer.apple.com/metal/jax/

step-by-step but it seems like JAX is still using CPU-only. At least, there is zero GPU utilization reported by `asitop`.

**Comment by kulinseth** [comment]
> Was anyone able to get that running? I followed
> 
> > https://developer.apple.com/metal/jax/
> 
> step-by-step but it seems like JAX is still using CPU-only. At least, there is zero GPU utilization reported by `asitop`. Tried that on an M2 Mac Mini with macOS 13.4.

@mlaves , can you send the `pip list` of the env ? Also which script were you trying ?

**Comment by Artur-Galstyan** [comment]
Cannot confirm that their little tutorial works. xla_backend still outputs CPU.

**Comment by martinkim0** [comment]
I was able to get jax-metal working following their tutorial. I get the following output when running `jax.devices()`:

```
Metal device set to: Apple M1

systemMemory: 16.00 GB
maxCacheSize: 5.33 GB

[MetalDevice(id=0, process_index=0)]
```

**Comment by godber** [comment]
Worked on M1 Max too

```
python my.py
Metal device set to: Apple M1 Max

systemMemory: 32.00 GB
maxCacheSize: 10.67 GB
```

**Comment by peregilk** [comment]
Seems to compile and install fine on M2 Pro. I think https://github.com/sanchit-gandhi/whisper-jax is a good example of the speed of Jax. This library is able to transcribe audio files with Whisper at a speed of 50-70x real time on a TPU v4, while the fastest PyTorch implementation is doing 3-4 x on a decent GPU. I was excited to test the speed on a Mac, but I am unable to get this to run. Ends with a "`failed to legalize operation 'mhlo.convolution'`" -error.

**Comment by mlaves** [comment]
> @mlaves , can you send the `pip list` of the env ? Also which script were you trying ?

@kulinseth The output of `pip list` is

```
Package    Version
---------- -------
jax        0.4.11
jax-metal  0.0.2
jaxlib     0.4.10
ml-dtypes  0.2.0
numpy      1.24.3
opt-einsum 3.3.0
pip        23.1.2
scipy      1.10.1
setuptools 67.6.1
six        1.16.0
wheel      0.40.0
```

I tried `python -c 'import jax; print(jax.devices())'`, which reports `[CpuDevice(id=0)]`.

**Comment by Artur-Galstyan** [comment]
As far as I can tell, there are some lax operations, which don't yet work properly with jax-metal. For instance `jax.lax.reduce_window` also results in the same `failed to legalize` error.

**Comment by MickaelRigault** [comment]
If it helps. For me (M1Max) the https://developer.apple.com/metal/jax/ tutorial did not work. 
Then, I updated my OS to Ventura (13.4), reran the tutorial (using conda env but that is the same) and it now works !
```
import jax
jax.devices()
>>> [MetalDevice(id=0, process_index=0)]
```

**Comment by hawkinsp** [comment]
I'm tempted to close this issue (support Metal GPUs: done!) and ask folks to file new issues for problems with the plugin, rather than trying to consolidate all known problems with the plugin into one gigantic github issue.

Sound good?

**Comment by skye** [comment]
Good idea, thanks Peter

BUT ONE MORE COMMENT @mlaves try updating `jaxlib` to 0.4.11 (not just jax). If you still don't see Metal support, please file a new issue :)

Also for anyone filing a Metal plugin issue, please feel free to tag me, @kulinseth, and @jyingl3. The issue should get triaged either way but that will help us look sooner. Thanks!

**Comment by nasyxx** [comment]
The Apple tutorial requires us to build jaxlib ourselves. 

Will pre-built wheels for the Metal support be provided in the future?

**Comment by hawkinsp** [comment]
@nasyxx Yes, the intention is that once the APIs stabilize a bit more, you should be able to use the prebuilt jaxlib wheels with the Metal plugin, with compatibility across a range of versions. Right now, it is safest to use *exactly* the version Apple is using.

**Comment by hawkinsp** [comment]
Closing: please open new issues for any problems with the JAX Metal plugin so we can look into and triage them separately.

It's exciting to see this land!

**Comment by kulinseth** [comment]
> If it helps. For me (M1Max) the https://developer.apple.com/metal/jax/ tutorial did not work. Then, I updated my OS to Ventura (13.4), reran the tutorial (using conda env but that is the same) and it now works !
> 
> ```
> import jax
> jax.devices()
> >>> [MetalDevice(id=0, process_index=0)]
> ```

Thanks for trying @MickaelRigault , JAX support is available in 13.4+ as mentioned on the webpage:
https://developer.apple.com/metal/jax/

**Comment by adam-hartshorne** [comment]
> > @mlaves , can you send the `pip list` of the env ? Also which script were you trying ?
> 
> @kulinseth The output of `pip list` is
> 
> ```
> Package    Version
> ---------- -------
> jax        0.4.11
> jax-metal  0.0.2
> jaxlib     0.4.10
> ml-dtypes  0.2.0
> numpy      1.24.3
> opt-einsum 3.3.0
> pip        23.1.2
> scipy      1.10.1
> setuptools 67.6.1
> six        1.16.0
> wheel      0.40.0
> ```
> 
> I tried `python -c 'import jax; print(jax.devices())'`, which reports `[CpuDevice(id=0)]`. I'm using python 3.11.
> 
> Update: After re-installing everything multiple times, I think the following did the trick: After following the steps in Apple's tutorial, I deactivated and reactivated the venv and now, everything works properly:
> 
> ```
> >>> python -c 'import jax; print(jax.devices())'
> Metal device set to: Apple M2 Pro
> 
> systemMemory: 16.00 GB
> maxCacheSize: 5.33 GB
> 
> [MetalDevice(id=0, process_index=0)]
> ```

I am having exactly the same issue on an M2 Macbook Air with jax-metal 0.0.3, but no amount of reinstalling / deactivating / reactivating the environment is fixing this. Exactly the same versions of all the other libraries and I have followed instructions on the website i.e. building the pinned jaxlib from source.
